ls1intum/Artemis
# Feature - Programming Exercises - New Create Dialog

### Checklist

### Motivation and Context


As described in https://github.com/ls1intum/Artemis/issues/801 the programming exercise generate dialog is not self-explanatory. It would be good to have more tooltips and in general a more user-friendly UI here. In particular, it would be good to rework the release/due date form dialogs into some sort of timeline, which displays all events in the lifecycle of an exercise and allows instructors to dynamically add dates (and tests)


### Description

There now is a new preview mode for the generation of a programming exercise, which displays the generated repository names and build plan IDs based on the short name of the course and exercise. This should help instructors to understand what is happening once a programming exercise is started and generated.

Furthermore, there now is a timeline, which replaces the date fields. You can see all important dates in the lifecycle of a programming exercise. Instructors can also dynamically add new dates (release/due/after due) and enable a new form of assessments: Manual reviews.
Enabling manual reviews will set the `assessmentType` in the exercise to `SEMI_AUTOMATIC`. In a follow-up PR, this will enable assessing programming exercises manually after all automatic tests have been run (As of now, there already is the option to manually overwrite a result, but this will then be enforced more strictly and only be possible if the "Manual Review" is enabled). 

I also added tooltips to the following labels:
* Title
* Short name
* Categories
* Timeline
* Timeline elements
* Preview
* Preview elements

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Create a new programming exercise
4. As soon as you type something in the short name field, there should be previews showing all to be generated repos and build plans
5. There are tooltips for all labels specified above
6. You can add and remove dates in the new timeline
7. You can enable "Manual Reviews" as soon as you have specified a due date
8. After saving the exercise and loading it again using "Edit" all fields are loaded with the values you specified when generating the exercise

### Screenshots


![preview_2 0](https://user-images.githubusercontent.com/27979674/66260090-55be9f00-e7ba-11e9-95fc-dced9188f7fb.gif)
